Client Shutdown is not handled correctly on Master

Steps to reproduce:
1. target/pack/bin/local -port 3000
2. target/pack/bin/word-count -master 127.0.0.1:3000

Wait for word-count to finish and note what the local process does. The akka disassociation messages are not handled well. The Executor Watcher should handle the disassociation messages within the Worker and the AppManager should reclaim idle workers where possible. Output is below:

[INFO] [11/07/2014 17:52:21.575] [Sum] Task TaskId(1,2) Throughput: (714368,5) (words, second)
[INFO] [11/07/2014 17:52:24.616] [AppManager] we get update UpdateSuccess(0,None)
[INFO] [11/07/2014 17:52:24.624] [ClockService] Application minClock tracking: Fri Nov 07 17:52:24 PST 2014
[INFO] [11/07/2014 17:52:26.561] [Sum] Task TaskId(1,1) Throughput: (1578539,5) (words, second)
[INFO] [11/07/2014 17:52:26.575] [Sum] Task TaskId(1,2) Throughput: (699322,5) (words, second)
[INFO] [11/07/2014 17:52:26.575] [Sum] Task TaskId(1,3) Throughput: (1617183,5) (words, second)
[INFO] [11/07/2014 17:52:27.735] [Master] Receive from client, Shutting down Application 0
[INFO] [11/07/2014 17:52:27.735] [AppManager] App Manager Shutting down application 0
[INFO] [11/07/2014 17:52:27.736] [AppManager] Shuttdown app master at akka://master/user/Worker1, appId: 0, executorId: -1
[INFO] [11/07/2014 17:52:27.741] [Worker-1] Worker shutting down executor: app0-executor-1 due to: AppMaster 0 shutdown requested by master...
[INFO] [11/07/2014 17:52:27.741] [Worker-1] Shuttting down child: akka://master/user/Worker1/app0-executor-1
[INFO] [11/07/2014 17:52:27.741] [AppManager] we get update UpdateSuccess(masterstate,None)
[INFO] [11/07/2014 17:52:27.743] [Worker] destroying executor process org.apache.gearpump.util.ActorSystemBooter
[INFO] [11/07/2014 17:52:27.750] [Worker-1] Executor is down app0-executor-1
[INFO] [11/07/2014 17:52:27.750] [Scheduler] Resource update id: 2002183369, slots: 96....
[INFO] [11/07/2014 17:52:28.007] [AppManager] Current elements: Set()
[INFO] [11/07/2014 17:52:28.020] [Sum] Task TaskId(1,0) Throughput: (6490517,10) (words, second)
[INFO] [11/07/2014 17:52:28.115] [Sum] Task TaskId(1,0) Throughput: (72824,0) (words, second)
[WARN] [11/07/2014 17:52:28.124] [master-akka.remote.default-remote-dispatcher-11] [akka.tcp://master@127.0.0.1:3000/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-1] Association with remote system [akka.tcp://app0-executor-1@192.168.1.11:51611] has failed, address is now gated for [5000] ms. Reason is: [Disassociated].
[WARN] [11/07/2014 17:52:28.146] [app0-executor0-akka.remote.default-remote-dispatcher-11] [akka.tcp://app0-executor0@192.168.1.11:51613/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-0] Association with remote system [akka.tcp://app0-executor-1@192.168.1.11:51611] has failed, address is now gated for [5000] ms. Reason is: [Disassociated].
[WARN] [11/07/2014 17:52:28.146] [app0-executor1-akka.remote.default-remote-dispatcher-5] [akka.tcp://app0-executor1@192.168.1.11:51614/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-0] Association with remote system [akka.tcp://app0-executor-1@192.168.1.11:51611] has failed, address is now gated for [5000] ms. Reason is: [Disassociated].
[INFO] [11/07/2014 17:52:28.151] [app0-executor0-akka.actor.default-dispatcher-17] [akka://app0-executor0/system/transports/akkaprotocolmanager.tcp0/akkaProtocol-tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-1] Message [akka.remote.transport.ActorTransportAdapter$DisassociateUnderlying] from Actor[akka://app0-executor0/deadLetters] to Actor[akka://app0-executor0/system/transports/akkaprotocolmanager.tcp0/akkaProtocol-tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-1#1215338712] was not delivered. [1] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:28.152] [app0-executor1-akka.actor.default-dispatcher-17] [akka://app0-executor1/system/transports/akkaprotocolmanager.tcp0/akkaProtocol-tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-1] Message [akka.remote.transport.ActorTransportAdapter$DisassociateUnderlying] from Actor[akka://app0-executor1/deadLetters] to Actor[akka://app0-executor1/system/transports/akkaprotocolmanager.tcp0/akkaProtocol-tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-1#-1474799343] was not delivered. [1] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:28.216] [app0-executor1-akka.actor.default-dispatcher-17] [akka://app0-executor1/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor1/system/remote-watcher#875963668] to Actor[akka://app0-executor1/deadLetters] was not delivered. [2] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:28.234] [app0-executor0-akka.actor.default-dispatcher-2] [akka://app0-executor0/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor0/system/remote-watcher#1485403649] to Actor[akka://app0-executor0/deadLetters] was not delivered. [2] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:28.261] [Sum] Task TaskId(1,0) Throughput: (99288,0) (words, second)
[INFO] [11/07/2014 17:52:28.376] [Sum] Task TaskId(1,0) Throughput: (91983,0) (words, second)
[INFO] [11/07/2014 17:52:29.203] [app0-executor0-akka.actor.default-dispatcher-4] [akka://app0-executor0/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor0/system/remote-watcher#1485403649] to Actor[akka://app0-executor0/deadLetters] was not delivered. [3] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:29.216] [app0-executor1-akka.actor.default-dispatcher-3] [akka://app0-executor1/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor1/system/remote-watcher#875963668] to Actor[akka://app0-executor1/deadLetters] was not delivered. [3] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.203] [app0-executor0-akka.actor.default-dispatcher-3] [akka://app0-executor0/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor0/system/remote-watcher#1485403649] to Actor[akka://app0-executor0/deadLetters] was not delivered. [4] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.216] [app0-executor1-akka.actor.default-dispatcher-17] [akka://app0-executor1/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor1/system/remote-watcher#875963668] to Actor[akka://app0-executor1/deadLetters] was not delivered. [4] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.748] [app0-executor0-akka.actor.default-dispatcher-17] [akka://app0-executor0/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor0/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/0/group_1_task_1#-1046615235] to Actor[akka://app0-executor0/deadLetters] was not delivered. [5] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.749] [app0-executor0-akka.actor.default-dispatcher-4] [akka://app0-executor0/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor0/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/0/group_0_task_0#-347974929] to Actor[akka://app0-executor0/deadLetters] was not delivered. [6] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.749] [app0-executor0-akka.actor.default-dispatcher-4] [akka://app0-executor0/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor0/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/0/group_0_task_1#-598577171] to Actor[akka://app0-executor0/deadLetters] was not delivered. [7] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.754] [app0-executor0-akka.actor.default-dispatcher-3] [akka://app0-executor0/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor0/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/0/group_1_task_0#260847679] to Actor[akka://app0-executor0/deadLetters] was not delivered. [8] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.756] [app0-executor1-akka.actor.default-dispatcher-17] [akka://app0-executor1/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor1/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/1/group_1_task_2#368072799] to Actor[akka://app0-executor1/deadLetters] was not delivered. [5] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.756] [app0-executor1-akka.actor.default-dispatcher-17] [akka://app0-executor1/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor1/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/1/group_1_task_3#1136396599] to Actor[akka://app0-executor1/deadLetters] was not delivered. [6] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.757] [app0-executor1-akka.actor.default-dispatcher-3] [akka://app0-executor1/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor1/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/1/group_0_task_3#1321548106] to Actor[akka://app0-executor1/deadLetters] was not delivered. [7] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:30.759] [app0-executor1-akka.actor.default-dispatcher-4] [akka://app0-executor1/deadLetters] Message [org.apache.gearpump.streaming.task.UpdateClock] from Actor[akka://app0-executor1/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/1/group_0_task_2#2108255720] to Actor[akka://app0-executor1/deadLetters] was not delivered. [8] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:31.203] [app0-executor0-akka.actor.default-dispatcher-17] [akka://app0-executor0/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor0/system/remote-watcher#1485403649] to Actor[akka://app0-executor0/deadLetters] was not delivered. [9] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:31.217] [app0-executor1-akka.actor.default-dispatcher-3] [akka://app0-executor1/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor1/system/remote-watcher#875963668] to Actor[akka://app0-executor1/deadLetters] was not delivered. [9] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:31.563] [Sum] Task TaskId(1,1) Throughput: (1882275,5) (words, second)
[INFO] [11/07/2014 17:52:31.576] [Sum] Task TaskId(1,0) Throughput: (2602450,3) (words, second)
[INFO] [11/07/2014 17:52:31.576] [Sum] Task TaskId(1,2) Throughput: (838574,5) (words, second)
[INFO] [11/07/2014 17:52:31.576] [Sum] Task TaskId(1,3) Throughput: (1939211,5) (words, second)
[INFO] [11/07/2014 17:52:32.201] [app0-executor0-akka.actor.default-dispatcher-17] [akka://app0-executor0/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor0/system/remote-watcher#1485403649] to Actor[akka://app0-executor0/deadLetters] was not delivered. [10] dead letters encountered, no more dead letters will be logged. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO] [11/07/2014 17:52:32.216] [app0-executor1-akka.actor.default-dispatcher-4] [akka://app0-executor1/deadLetters] Message [akka.remote.RemoteWatcher$Heartbeat$] from Actor[akka://app0-executor1/system/remote-watcher#875963668] to Actor[akka://app0-executor1/deadLetters] was not delivered. [10] dead letters encountered, no more dead letters will be logged. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[WARN] [11/07/2014 17:52:33.228] [app0-executor1-akka.remote.default-remote-dispatcher-5] [akka.tcp://app0-executor1@192.168.1.11:51614/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-0/endpointWriter] AssociationError [akka.tcp://app0-executor1@192.168.1.11:51614] -> [akka.tcp://app0-executor-1@192.168.1.11:51611]: Error [Invalid address: akka.tcp://app0-executor-1@192.168.1.11:51611] [
akka.remote.InvalidAssociation: Invalid address: akka.tcp://app0-executor-1@192.168.1.11:51611
Caused by: akka.remote.transport.Transport$InvalidAssociationException: Connection refused: /192.168.1.11:51611
]
[WARN] [11/07/2014 17:52:33.234] [app0-executor1-akka.remote.default-remote-dispatcher-5] [Remoting] Tried to associate with unreachable remote address [akka.tcp://app0-executor-1@192.168.1.11:51611]. Address is now gated for 5000 ms, all messages to this address will be delivered to dead letters. Reason: Connection refused: /192.168.1.11:51611
[INFO] [11/07/2014 17:52:33.244] [ActorSystemBooter] System app0-executor1 Watched actor is terminated Actor[akka://app0-executor1/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/1#500300830]
[INFO] [11/07/2014 17:52:33.245] [ActorSystemBooter] Actor System app0-executor1 is shutting down...
[INFO] [11/07/2014 17:52:33.245] [Client] closing netty client netty-client-HostPort(192.168.1.11,51619)...
[INFO] [11/07/2014 17:52:33.251] [app0-executor1-akka.remote.default-remote-dispatcher-5] [akka.tcp://app0-executor1@192.168.1.11:51614/system/remoting-terminator] Shutting down remote daemon.
[INFO] [11/07/2014 17:52:33.253] [app0-executor1-akka.remote.default-remote-dispatcher-5] [akka.tcp://app0-executor1@192.168.1.11:51614/system/remoting-terminator] Remote daemon shut down; proceeding with flushing remote transports.
[WARN] [11/07/2014 17:52:33.257] [app0-executor0-akka.remote.default-remote-dispatcher-11] [akka.tcp://app0-executor0@192.168.1.11:51613/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fapp0-executor-1%40192.168.1.11%3A51611-0/endpointWriter] AssociationError [akka.tcp://app0-executor0@192.168.1.11:51613] -> [akka.tcp://app0-executor-1@192.168.1.11:51611]: Error [Invalid address: akka.tcp://app0-executor-1@192.168.1.11:51611] [
akka.remote.InvalidAssociation: Invalid address: akka.tcp://app0-executor-1@192.168.1.11:51611
Caused by: akka.remote.transport.Transport$InvalidAssociationException: Connection refused: /192.168.1.11:51611
]
[ERROR] [11/07/2014 17:52:33.256] [Client] Connection failed netty-client-HostPort(192.168.1.11,51620)
java.nio.channels.ClosedChannelException
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.cleanUpWriteBuffer(AbstractNioWorker.java:433)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.writeFromUserCode(AbstractNioWorker.java:128)
    at org.jboss.netty.channel.socket.nio.NioClientSocketPipelineSink.eventSunk(NioClientSocketPipelineSink.java:84)
    at org.jboss.netty.channel.Channels.write(Channels.java:725)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.doEncode(OneToOneEncoder.java:71)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:59)
    at org.jboss.netty.channel.Channels.write(Channels.java:704)
    at org.jboss.netty.channel.Channels.write(Channels.java:671)
    at org.jboss.netty.channel.AbstractChannel.write(AbstractChannel.java:248)
    at org.apache.gearpump.transport.netty.Client.flushRequest(Client.scala:148)
    at org.apache.gearpump.transport.netty.Client.org$apache$gearpump$transport$netty$Client$$send(Client.scala:130)
    at org.apache.gearpump.transport.netty.Client$$anonfun$messageHandler$1.applyOrElse(Client.scala:61)
    at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:165)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
    at org.apache.gearpump.transport.netty.Client.aroundReceive(Client.scala:35)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
    at akka.dispatch.Mailbox.run(Mailbox.scala:220)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[INFO] [11/07/2014 17:52:33.256] [Client] failed to send requests to /192.168.1.11:51620
java.nio.channels.ClosedChannelException
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.cleanUpWriteBuffer(AbstractNioWorker.java:433)
    at org.jboss.netty.channel.socket.nio.AbstractNioWorker.writeFromUserCode(AbstractNioWorker.java:128)
    at org.jboss.netty.channel.socket.nio.NioClientSocketPipelineSink.eventSunk(NioClientSocketPipelineSink.java:84)
    at org.jboss.netty.channel.Channels.write(Channels.java:725)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.doEncode(OneToOneEncoder.java:71)
    at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:59)
    at org.jboss.netty.channel.Channels.write(Channels.java:704)
    at org.jboss.netty.channel.Channels.write(Channels.java:671)
    at org.jboss.netty.channel.AbstractChannel.write(AbstractChannel.java:248)
    at org.apache.gearpump.transport.netty.Client.flushRequest(Client.scala:148)
    at org.apache.gearpump.transport.netty.Client.org$apache$gearpump$transport$netty$Client$$send(Client.scala:130)
    at org.apache.gearpump.transport.netty.Client$$anonfun$messageHandler$1.applyOrElse(Client.scala:61)
    at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:165)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
    at org.apache.gearpump.transport.netty.Client.aroundReceive(Client.scala:35)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
    at akka.dispatch.Mailbox.run(Mailbox.scala:220)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[ERROR] [11/07/2014 17:52:33.264] [app0-executor0-akka.actor.default-dispatcher-4] [akka://app0-executor0/user/$a] null
java.lang.NullPointerException
    at org.apache.gearpump.transport.netty.Client$$anonfun$messageHandler$1.applyOrElse(Client.scala:60)
    at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:165)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
    at org.apache.gearpump.transport.netty.Client.aroundReceive(Client.scala:35)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
    at akka.dispatch.Mailbox.run(Mailbox.scala:220)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[INFO] [11/07/2014 17:52:33.266] [Client] closing netty client netty-client-HostPort(192.168.1.11,51620)...
[WARN] [11/07/2014 17:52:33.267] [app0-executor0-akka.remote.default-remote-dispatcher-11] [Remoting] Tried to associate with unreachable remote address [akka.tcp://app0-executor-1@192.168.1.11:51611]. Address is now gated for 5000 ms, all messages to this address will be delivered to dead letters. Reason: Connection refused: /192.168.1.11:51611
[INFO] [11/07/2014 17:52:33.271] [Client] netty client connect to netty-client-HostPort(192.168.1.11,51620), tries: 0， hostPort: HostPort(192.168.1.11,51620)
[INFO] [11/07/2014 17:52:33.274] [Client] netty client connect to netty-client-HostPort(192.168.1.11,51620), tries: 0， hostPort: HostPort(192.168.1.11,51620)
[INFO] [11/07/2014 17:52:33.298] [ForkJoinPool-3-worker-15] [Remoting] Remoting shut down
[INFO] [11/07/2014 17:52:33.299] [app0-executor1-akka.remote.default-remote-dispatcher-5] [akka.tcp://app0-executor1@192.168.1.11:51614/system/remoting-terminator] Remoting shut down.
[INFO] [11/07/2014 17:52:33.304] [Context] Context.term, cleanup resources...., we have 3 items to close...
[INFO] [11/07/2014 17:52:33.305] [Context] closing Client actor....
[INFO] [11/07/2014 17:52:33.305] [Context] Closing all client resources....
[INFO] [11/07/2014 17:52:33.306] [Context] Closing all server resources....
[INFO] [11/07/2014 17:52:33.343] [Worker] Executor exit normally with exit value 0
[INFO] [11/07/2014 17:52:33.344] [Worker-1] Executor is down app0-executor1
[INFO] [11/07/2014 17:52:33.344] [Scheduler] Resource update id: -1031077879, slots: 100....
[INFO] [11/07/2014 17:52:33.361] [ActorSystemBooter] System app0-executor0 Watched actor is terminated Actor[akka://app0-executor0/remote/akka.tcp/app0-executor-1@192.168.1.11:51611/user/daemon/appmaster/0#-1862698346]
[INFO] [11/07/2014 17:52:33.361] [ActorSystemBooter] Actor System app0-executor0 is shutting down...
[INFO] [11/07/2014 17:52:33.362] [Client] closing netty client netty-client-HostPort(192.168.1.11,51620)...
[INFO] [11/07/2014 17:52:33.367] [app0-executor0-akka.remote.default-remote-dispatcher-11] [akka.tcp://app0-executor0@192.168.1.11:51613/system/remoting-terminator] Shutting down remote daemon.
[INFO] [11/07/2014 17:52:33.368] [app0-executor0-akka.remote.default-remote-dispatcher-11] [akka.tcp://app0-executor0@192.168.1.11:51613/system/remoting-terminator] Remote daemon shut down; proceeding with flushing remote transports.
[INFO] [11/07/2014 17:52:33.399] [ForkJoinPool-3-worker-15] [Remoting] Remoting shut down
[INFO] [11/07/2014 17:52:33.400] [app0-executor0-akka.remote.default-remote-dispatcher-11] [akka.tcp://app0-executor0@192.168.1.11:51613/system/remoting-terminator] Remoting shut down.
[INFO] [11/07/2014 17:52:33.405] [Context] Context.term, cleanup resources...., we have 3 items to close...
[INFO] [11/07/2014 17:52:33.406] [Context] closing Client actor....
[INFO] [11/07/2014 17:52:33.406] [Context] Closing all client resources....
[INFO] [11/07/2014 17:52:33.408] [Context] Closing all server resources....
[INFO] [11/07/2014 17:52:33.458] [Worker] Executor exit normally with exit value 0
[INFO] [11/07/2014 17:52:33.458] [Worker-1] Executor is down app0-executor0
[INFO] [11/07/2014 17:52:33.458] [Scheduler] Resource update id: 2002183369, slots: 100....
[WARN] [11/07/2014 17:52:33.490] [master-akka.remote.default-remote-dispatcher-11] [Remoting] Tried to associate with unreachable remote address [akka.tcp://app0-executor-1@192.168.1.11:51611]. Address is now gated for 5000 ms, all messages to this address will be delivered to dead letters. Reason: Connection refused: /192.168.1.11:51611
[INFO] [11/07/2014 17:52:33.490] [AppManager] App Master is terminiated, network down: true

--- a/README.md
+++ b/README.md
@@ -53,14 +53,14 @@ Per initial benchmarks we are able to process 2 million messages/second (100 byt
   core/target/pack/bin/local -port 3000
   ```
 
-2. Start WordCount or SOL Example
+2. Start an Example
   
   ```bash
   ## Run WordCount example
   target/pack/bin/gear app -jar ./examples/wordcount/target/pack/lib/gearpump-examples-wordcount-0.2-SNAPSHOT.jar org.apache.gearpump.streaming.examples.wordcount.WordCount -master 127.0.0.1:3000
   ```
   ```bash
-  ## Run SOL examples
+  ## Run SOL example
   target/pack/bin/gear app -jar ./examples/sol/target/pack/lib/gearpump-examples-sol-0.2-SNAPSHOT.jar org.apache.gearpump.streaming.examples.sol.SOL -master 127.0.0.1:3000
   ```
   ```bash
@@ -98,7 +98,7 @@ Per initial benchmarks we are able to process 2 million messages/second (100 byt
   target/pack/bin/worker
   ```
   
-4. Start Client Example Code
+4. Start a Client Example 
   ```bash
   ## Run WordCount example
   target/pack/bin/gear app -jar ./examples/wordcount/target/pack/lib/gearmp-examples-wordcount-0.2-SNAPSHOT.jar org.apache.gearpump.streaming.examples.wordcount.WordCount -master 127.0.0.1:3000

--- a/core/src/main/resources/reference.conf
+++ b/core/src/main/resources/reference.conf
@@ -69,24 +69,29 @@ master {
     "akka.contrib.datareplication.DataReplication$"
   ]
   akka {
+    loglevel = "INFO"
+    log-dead-letters = off
+    log-dead-letters-during-shutdown = off
     actor {
       ## Master forms a akka cluster
       provider = "akka.cluster.ClusterActorRefProvider"
-    }
-    remote {
-      log-remote-lifecycle-events = off
     }
     cluster {
       roles = ["master"]
       auto-down-unreachable-after = 15s
     }
-    log-dead-letters = off
+    remote {
+      log-remote-lifecycle-events = off
+    }
   }
 }
 
 worker {
   ## Add worker overrided config
   akka {
+    loglevel = "INFO"
+    log-dead-letters = off
+    log-dead-letters-during-shutdown = off
     actor {
       provider = "akka.remote.RemoteActorRefProvider"
     }
@@ -95,6 +100,9 @@ worker {
       resource {
         slots = 100
       }
+    }
+    remote {
+      log-remote-lifecycle-events = off
     }
   }
 }
@@ -106,6 +114,8 @@ akka {
     "org.apache.gearpump.metrics.Metrics$"
   ]
   loglevel = "INFO"
+  log-dead-letters = off
+  log-dead-letters-during-shutdown = off
   actor {
     provider = "akka.remote.RemoteActorRefProvider"
     default-mailbox {
@@ -150,7 +160,7 @@ akka {
     }
   }
   remote {
-#    use-dispatcher = ""
+    log-remote-lifecycle-events = off
     use-dispatcher = "akka.remote.default-remote-dispatcher"
     enabled-transports = ["akka.remote.netty.tcp"]
     netty.tcp {

--- a/core/src/main/scala/org/apache/gearpump/cluster/Master.scala
+++ b/core/src/main/scala/org/apache/gearpump/cluster/Master.scala
@@ -19,6 +19,7 @@
 package org.apache.gearpump.cluster
 
 import akka.actor._
+import akka.remote.DisassociatedEvent
 import com.typesafe.config.Config
 import org.apache.gearpump.cluster.AppMasterToMaster._
 import org.apache.gearpump.cluster.ClientToMaster._
@@ -49,7 +50,12 @@ private[cluster] class Master extends Actor with Stash {
 
   LOG.info("master is started at " + ActorUtil.getFullPath(context) + "...")
 
-  override def receive : Receive = workerMsgHandler orElse appMasterMsgHandler orElse clientMsgHandler orElse terminationWatch orElse ActorUtil.defaultMsgHandler(self)
+  override def receive : Receive = workerMsgHandler orElse
+    appMasterMsgHandler orElse
+    clientMsgHandler orElse
+    terminationWatch orElse
+    disassociated orElse
+    ActorUtil.defaultMsgHandler(self)
 
   final val undefinedUid = 0
   @tailrec final def newUid(): Int = {
@@ -105,6 +111,12 @@ private[cluster] class Master extends Actor with Stash {
       appManager.forward(app)
   }
 
+  def disassociated : Receive = {
+    case disassociated : DisassociatedEvent =>
+      LOG.info(s" disassociated ${disassociated.remoteAddress}")
+      //LOG.info(s"remote lifecycle events are "+systemConfig.getString("akka.remote.log-remote-lifecycle-events"))
+  }
+
   def terminationWatch : Receive = {
     case t : Terminated =>
       val actor = t.actor
@@ -123,6 +135,7 @@ private[cluster] class Master extends Actor with Stash {
     val schedulerClass = Class.forName(systemConfig.getString(Constants.GEARPUMP_SCHEDULING_SCHEDULER))
     appManager = context.actorOf(Props[AppManager], classOf[AppManager].getSimpleName)
     scheduler = context.actorOf(Props(schedulerClass))
+    context.system.eventStream.subscribe(self, classOf[DisassociatedEvent]);
   }
 }
 

--- a/core/src/main/scala/org/apache/gearpump/cluster/main/Local.scala
+++ b/core/src/main/scala/org/apache/gearpump/cluster/main/Local.scala
@@ -18,9 +18,8 @@
 
 package org.apache.gearpump.cluster.main
 
-import akka.actor.{ActorSystem, Props}
+import akka.actor.{Props, ActorSystem}
 import com.typesafe.config.ConfigValueFactory
-import org.apache.gearpump.cluster.Master
 import org.apache.gearpump.util.Constants._
 import org.apache.gearpump.util.{ActorUtil, Configs}
 import org.slf4j.{Logger, LoggerFactory}
@@ -54,7 +53,7 @@ object Local extends App with ArgumentsParser {
       withValue("akka.remote.netty.tcp.hostname", ConfigValueFactory.fromAnyRef(ip)).
       withValue("gearpump.cluster.masters",  ConfigValueFactory.fromAnyRef(List(s"$ip:$port").asJava))
     )
-    val master = system.actorOf(Props[Master], MASTER)
+    val master = system.actorOf(Props[org.apache.gearpump.cluster.Master], MASTER)
     val masterPath = ActorUtil.getSystemPath(system) + s"/user/$MASTER"
     LOG.info(s"master is started at $masterPath...")
 

--- a/core/src/main/scala/org/apache/gearpump/cluster/main/Worker.scala
+++ b/core/src/main/scala/org/apache/gearpump/cluster/main/Worker.scala
@@ -19,7 +19,7 @@
 package org.apache.gearpump.cluster.main
 
 import akka.actor.{ActorSystem, Props}
-import org.apache.gearpump.cluster.{MasterProxy, Worker}
+import org.apache.gearpump.cluster.MasterProxy
 import org.apache.gearpump.transport.HostPort
 import org.apache.gearpump.util.Configs
 import org.apache.gearpump.util.Constants._
@@ -34,7 +34,7 @@ object Worker extends App with ArgumentsParser {
 
   val options = Array.empty[(String, CLIOption[Any])]
 
-  def start() = {
+  def start(): Unit = {
     worker()
   }
 
@@ -50,9 +50,10 @@ object Worker extends App with ArgumentsParser {
     }
     val masterProxy = system.actorOf(Props(classOf[MasterProxy], masterAddress), MASTER)
 
-    val worker = system.actorOf(Props(classOf[Worker], masterProxy), classOf[Worker].getSimpleName + id)
+    system.actorOf(Props(classOf[org.apache.gearpump.cluster.Worker], masterProxy),
+      classOf[org.apache.gearpump.cluster.Worker].getSimpleName + id)
 
-    system.awaitTermination
+    system.awaitTermination()
   }
 
   start()

--- a/project/Build.scala
+++ b/project/Build.scala
@@ -16,9 +16,13 @@ object Build extends sbt.Build {
   val akkaVersion = "2.3.6"
   val kryoVersion = "0.3.2"
   val codahaleVersion = "3.0.2"
-  val commonsLangVersion = "3.3.2"
+  val commonsCodecVersion = "1.6"
   val commonsHttpVersion = "3.1"
+  val commonsLangVersion = "3.3.2"
+  val commonsLoggingVersion = "1.1.3"
+  val findbugsVersion = "2.0.1"
   val gearPumpVersion = "0.2-SNAPSHOT"
+  val guavaVersion = "15.0"
   val dataReplicationVersion = "0.7"
   val hadoopVersion = "2.5.1"
   val jgraphtVersion = "0.9.0"
@@ -85,8 +89,12 @@ object Build extends sbt.Build {
         "org.slf4j" % "jul-to-slf4j" % slf4jVersion,
         "org.slf4j" % "jcl-over-slf4j" % slf4jVersion,
         "org.fusesource" % "sigar" % sigarVersion classifier("native"),
+        "com.google.code.findbugs" % "jsr305" % findbugsVersion,
         "org.apache.commons" % "commons-lang3" % commonsLangVersion,
+        "commons-logging" % "commons-logging" % commonsLoggingVersion,
         "commons-httpclient" % "commons-httpclient" % commonsHttpVersion,
+        "commons-codec" % "commons-codec" % commonsCodecVersion,
+        "com.google.guava" % "guava" % guavaVersion,
         "com.typesafe.akka" %% "akka-actor" % akkaVersion,
         "com.typesafe.akka" %% "akka-remote" % akkaVersion,
         "com.typesafe.akka" %% "akka-cluster" % akkaVersion,

--- a/streaming/src/main/resources/reference.conf
+++ b/streaming/src/main/resources/reference.conf
@@ -4,12 +4,26 @@ gearpump {
       vmargs = "-server -Xms512M -Xmx1024M -Xss1M -XX:MaxPermSize=128m -XX:+HeapDumpOnOutOfMemoryError -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=80 -XX:+UseParNewGC -XX:NewRatio=3"
       extraClasspath = ""
 ##      vmargs = "-server -Xms1024M -Xmx4096M -Xss1M -XX:MaxPermSize=128m -XX:+HeapDumpOnOutOfMemoryError -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=80 -XX:+UseParNewGC -XX:NewRatio=3 -XX:NewSize=512m"
-	}
+      akka {
+        log-dead-letters = off
+        log-dead-letters-during-shutdown = off
+        remote {
+          log-remote-lifecycle-events = off
+        }
+      }
+    }
 
     executor {
       vmargs = "-server -Xms512M -Xmx1024M -Xss1M -XX:MaxPermSize=128m -XX:+HeapDumpOnOutOfMemoryError -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=80 -XX:+UseParNewGC -XX:NewRatio=3"
       extraClasspath = ""
 #      vmargs = "-server -Xms1024M -Xmx4096M -Xss1M -XX:MaxPermSize=128m -XX:+HeapDumpOnOutOfMemoryError  -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=80 -XX:+UseParNewGC -XX:NewRatio=3 -XX:NewSize=512m"
+      akka {
+        log-dead-letters = off
+        log-dead-letters-during-shutdown = off
+        remote {
+          log-remote-lifecycle-events = off
+        }
+      }
     }
   }
 }

